presto-main/src/main/java/com/facebook/presto/execution/buffer/SpoolingOutputBuffer.java

```diff
@@ -0,0 +1,788 @@
+            }
+            else {
+                pagesAcknowledged += sequenceId - range.lowerEndpoint() - startPage.get();
+                startPage.set(toIntExact(sequenceId - range.lowerEndpoint()));
+                break;
+            }
+        }
+
+        totalBufferedPages.addAndGet(-pagesRemoved);
+        totalBufferedBytes.addAndGet(-bytesRemoved);
+        totalPagesRemaining.addAndGet(-pagesAcknowledged);
+
+        return pagesAcknowledged;
+    }
+
+    private synchronized void acknowledgePagesFromMemory(long sequenceId, long startingSequenceId)
+    {
+        checkState(startingSequenceId >= currentMemorySequenceId.get(), "Invalid startingSequenceId for memory pages");
+        int pagesToRemove = toIntExact(sequenceId - startingSequenceId);
+        checkArgument(pagesToRemove <= pages.size(), "Invalid sequenceId");
+
+        long bytesRemoved = 0;
+        for (int i = 0; i < pagesToRemove; i++) {
+            SerializedPage removedPage = pages.remove();
+            bytesRemoved += removedPage.getRetainedSizeInBytes();
+            currentMemorySequenceId.incrementAndGet();
+        }
+
+        totalBufferedPages.addAndGet(-pagesToRemove);
+        totalBufferedBytes.addAndGet(-bytesRemoved);
+        totalPagesRemaining.addAndGet(-pagesToRemove);
+    }
+
+    @Override
+    public void abort(OutputBufferId bufferId)
+    {
+        checkArgument(bufferId.getId() == outputBufferId.getId(), "Invalid outputBufferId");
+        destroy();
+    }
+
+    @Override
+    public void setNoMorePages()
+    {
+        PendingRead pendingRead;
+        synchronized (this) {
+            state.compareAndSet(OPEN, NO_MORE_PAGES);
+            state.compareAndSet(NO_MORE_BUFFERS, FLUSHING);
+            noMorePages.set(true);
+
+            pendingRead = this.pendingRead;
+            this.pendingRead = null;
+
+            log.info("A total of %s pages and %s bytes was written into TempStorage", totalStoragePagesAdded.get(), totalStorageBytesAdded.get());
+        }
+
+        if (pendingRead != null) {
+            processPendingRead(pendingRead);
+        }
+
+        checkFlushComplete();
+    }
+
+    private void checkFlushComplete()
+    {
+        if (state.get() != FLUSHING) {
+            return;
+        }
+
+        if (totalBufferedPages.get() == 0) {
+            destroy();
+        }
+    }
+
+    @Override
+    public void destroy()
+    {
+        PendingRead pendingRead;
+        synchronized (this) {
+            if (state.setIf(FINISHED, oldState -> !oldState.isTerminal())) {
+                close();
+            }
+            pendingRead = this.pendingRead;
+            this.pendingRead = null;
+        }
+
+        if (pendingRead != null) {
+            pendingRead.completeResultFutureWithEmpty();
+        }
+    }
+
+    @Override
+    public void fail()
+    {
+        if (state.setIf(BufferState.FAILED, oldState -> !oldState.isTerminal())) {
+            finalizerService.addFinalizer(this, this::close);
```

Comment:
We should make sure this is called no matter what, not just when the buffer goes to failed state.